torch/_inductor/memory.py

```diff
@@ -0,0 +1,555 @@
+    """
+    from .scheduler import SchedulerBuffer
+
+    for node in nodes:
+        node.mpi_node.pred_nodes = list(
+            {
+                name_to_fused_node[pred_buffer.defining_op.get_name()]
+                for pred_buffer in node.mpi_node.pred_buffers
+                if (
+                    isinstance(pred_buffer, SchedulerBuffer)
+                    and pred_buffer.defining_op.get_name() in name_to_fused_node
+                )
+            }
+        )
+        node.mpi_node.succ_nodes = list(
+            {
+                succ_node
+                for buffer in node.get_outputs()
+                for succ_node in buffer.mpi_buffer.succ_nodes
+            }
+        )
+
+
+def topological_sort_lpmf(
+    nodes: List[BaseSchedulerNode],
+    name_to_input_buf: Dict[str, FreeableInputBuffer],
+    name_to_buf: Dict[str, SchedulerBuffer],
+    graph_outputs: Set[str],
+) -> List[BaseSchedulerNode]:
+    """
+    A bfs-based greedy topological order. LPMF stands for "Least Peak Memory First".
+    The algorithm maintain the max memory so far.
+    At every iteration, for each scheduleable node, it computes:
+        - how much memory needs to be allocated for the output buffers of this node;
+        - how much memory can be freed as a result of executing this node.
+    This gives us two values for each node:
+        (1) mem1: memory during the execution of the node;
+        (2) mem2: memory after executing the node, after some input buffers are freed.
+    The greedy approach select as follows:
+        (i) if there are nodes whose mem1 values are below the max memory so far,
+            then pick the node with the lowest mem2 value;
+        (ii) otherwise, pick the one with the lowest mem1 value.
+    """
+
+    # compute nodes' number of unmet dependencies (for schedulability)
+    # initialize the list of nodes ready to be scheduled
+    nodes_to_schedule: Set[BaseSchedulerNode] = set()
+    for node in nodes:
+        # note that .unmet_dependencies could have deps with the same name
+        # and in that case, it should only be counted once
+        node.mpi_node.indegree = len(node.mpi_node.pred_nodes)
+        if node.mpi_node.indegree == 0:
+            nodes_to_schedule.add(node)
+
+    # compute buffers' number of unmet successors (used to decide when to free)
+    for buf in list(name_to_buf.values()) + list(name_to_input_buf.values()):
+        buf.mpi_buffer.outdegree = len(buf.mpi_buffer.succ_nodes)
+        if buf.get_name() in graph_outputs:
+            buf.mpi_buffer.outdegree += 1
+
+    # initialize memory estimations
+    live_memory = sum(
+        input_buf.mpi_buffer.size_free for input_buf in name_to_input_buf.values()
+    )
+
+    # this is the total output memory, which is a lower bound for peak memory
+    output_memory = sum(
+        name_to_buf[buf_name].mpi_buffer.size_free
+        for buf_name in graph_outputs
+        if buf_name in name_to_buf
```

Comment:
when is this not set ?

Reply:
Good question. For the forward pass, `graph_outputs` can have things like `primals_*` but won't be in `name_to_buf` because they are not schedulerNodes' output buffers.